src/actions/attack.cpp

```diff
@@ -1604,6 +1604,63 @@ int under_leadership(const unit &u, const map_location& loc, const_attack_ptr we
 	return leader_effect.get_composite_value();
 }
 
+int attack_type::specials_alignment(int combat_modifier, bool is_fearless) const
+{
+	if(!has_special_or_ability("special_alignment")){
+		return combat_modifier;
+	}
+	// if "alignment" attribute not eqauls to string in checking_alignment chain,
+	// or not "special_alignment special/abilities then alignment_type returned by default.
+	const std::set<std::string> checking_alignment = {"neutral", "lawful", "chaotic", "liminal"};
+	unit_ability_list abil_list = get_specials_and_abilities("special_alignment");
+	for(unit_ability_list::iterator i = abil_list.begin(); i != abil_list.end();) {
+		if(checking_alignment.count((*i->ability_cfg)["alignment"].str()) == 0){
+			i = abil_list.erase(i);
```

Comment:
IIUC, this code would allow a weapon to have a special that made the attack `lawful`, and also a special that made the same attack `chaotic`. This code would allow both of those through, and later would calculate the `lawful_value` and `chaotic_value`, and finally would add them together at which time they would cancel each other out.

For specials that change the damage type, for example the Holy Water item, I think it just uses "last one added wins". Following that pattern, if a unit picked up "make your melee attacks lawful" followed by "make your melee attacks chaotic", I'd expect the melee attacks to get the chaotic bonus.

I think that could simplify the code too - it could use the `for` loop's ranged syntax because it wouldn't need to erase items from the list, just store the last-seen alignment. Some of the later code might then be replaced with a call to `generic_combat_modifier()`.

Reply:
If many specials presnet how choose alignment used (i think in 'leadership' cases)?

Reply:
Good question. I guess it could happen for leadership-style changes of the damage type too, if there was a Mage of Light with "bless all adjacent units' weapons (change to `arcane` damage)" and another with "all adjacent units' weapons ignore armour (change to Era of Magic's `secret` damage type)".